### DISCUSSION TITLE
refresh_token is required in either the JSON payload or Cookie

### DISCUSSION THREAD

**Discussion by Pachat** [directus/directus#15589] [Archive]
Hi,
I get this error and it appears only when my test is finishing.
![Directus](https://user-images.githubusercontent.com/13746133/190370886-e3d5d737-8b77-4404-8611-a6550147bc8a.png)

Console gives
```
Event:                    request cypress_runner.js:201254:13
Resource type:            xhr cypress_runner.js:201254:13
Method:                   POST cypress_runner.js:201254:13
Url:                      http://localhost:8023/auth/refresh cypress_runner.js:201254:13
Request went to origin?:  yes cypress_runner.js:201254:13
Xhr:                      
XMLHttpRequest { method: "POST", url: "http://localhost:8023/auth/refresh", id: "xhr62", canceled: true, onreadystatechange: null, readyState: 0, timeout: 0, withCredentials: true, upload: XMLHttpRequestUpload, responseURL: "", … }
cypress_runner.js:201254:13
Request headers:          
Object { host: "localhost:8023", "user-agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:104.0) Gecko/20100101 Firefox/104.0", accept: "application/json, text/plain, */*", "accept-language": "fr,fr-FR;q=0.8,en-US;q=0.5,en;q=0.3", "accept-encoding": "gzip, deflate", referer: "http://localhost:8023/admin/login", "cache-control": "no-store", origin: "http://localhost:8023", connection: "keep-alive" }
cypress_runner.js:201254:13
Response status code:     400 cypress_runner.js:201254:13
Response headers:         
Object { "x-powered-by": "Directus", "access-control-allow-origin": "http://localhost:8023", vary: "Origin, Accept-Encoding", "access-control-allow-credentials": "true", "access-control-expose-headers": "Content-Range", "content-type": "application/json; charset=utf-8", etag: "W/\"85-eTA01tQQkdFQTnPlqwL4x3RG/8Y\"", date: "Thu, 15 Sep 2022 09:18:45 GMT", "keep-alive": "timeout=5", "content-encoding": "gzip", … }
cypress_runner.js:201254:13
Response body:            {"errors":[{"message":"\"refresh_token\" is required in either the JSON payload or Cookie","extensions":{"code":"INVALID_PAYLOAD"}}]}

```
and my `.env` is
```
...
ACCESS_TOKEN_TTL="15m"
REFRESH_TOKEN_TTL="7d"
REFRESH_TOKEN_COOKIE_SECURE=false
REFRESH_TOKEN_COOKIE_SAME_SITE="lax"
REFRESH_TOKEN_COOKIE_NAME="direct_23_refresh_token"
REFRESH_TOKEN_COOKIE_DOMAIN="http://localhost:8023"
...

```

How can I avoid this error?

**Comment by br41nslug** [comment]
How are you passing the `refresh_token`? If you're using a JSON payload do you have the `mode` set to `json` in the payload?

**Reply by Pachat** [reply]
Thank you for the suggestion.
I'm only visiting pages, and the `/auth/refresh` is called by Directus I don't know where from.

I made some later tries with the test tool I use, Cypress, and found out that 

- if I call the page directly by the URL, as I was doing it,  `cy.visit('/admin/content/myurl')`, I get the error. But
- if I call the page by simulating a click in the menu `cy.get('[href="/admin/content/myurl"]').click()`, the error disappears.

Strange? But as the problem disappears, I'll leave it with that solution.

**Reply by br41nslug** [reply]
Ah sorry my mistake, haven't worked with cyprus personally. My guess would be that cyprus needs some configuration to properly deal with cookies perhaps 🤔 I remember that being a thing with old test frameworks like Selenium or TestCafe.

> I'm only visiting pages, and the /auth/refresh is called by Directus I don't know where from.

The auth/refresh is called on page load to check if the saved credentials are still valid and redirect to login if not. This is probably not happening when simulating the click because the page doesnt refresh for that.
